File: go/internal/storage/markdown/sections/business_model.go
package sections

import (
	"fmt"
	"regexp"
	"strings"
	"time"

	"github.com/studiowebux/mdplanner/internal/domain"
)

// BusinessModelParser handles parsing and serializing Business Model Canvases from markdown
type BusinessModelParser struct {
	idPattern  *regexp.Regexp
	sectionMap map[string]string
}

// NewBusinessModelParser creates a new Business Model Canvas parser
func NewBusinessModelParser() *BusinessModelParser {
	return &BusinessModelParser{
		idPattern: regexp.MustCompile(`<!-- id: ([^ ]+)`),
		sectionMap: map[string]string{
			"### Key Partners":           "keyPartners",
			"### Key Activities":         "keyActivities",
			"### Key Resources":          "keyResources",
			"### Value Proposition":      "valueProposition",
			"### Customer Relationships": "customerRelationships",
			"### Channels":               "channels",
			"### Customer Segments":      "customerSegments",
			"### Cost Structure":         "costStructure",
			"### Revenue Streams":        "revenueStreams",
		},
	}
}

// Parse extracts Business Model Canvases from lines
// Assumes lines are already extracted for this section
func (p *BusinessModelParser) Parse(lines []string) []domain.BusinessModelCanvas {
	var canvases []domain.BusinessModelCanvas
	var current *domain.BusinessModelCanvas
	var currentSubsection string

	for i := 0; i < len(lines); i++ {
		line := lines[i]
		trimmed := strings.TrimSpace(line)

		// Skip section headers if present
		if strings.HasPrefix(trimmed, "# Business Model Canvas") || trimmed == "<!-- Business Model Canvas -->" {
			continue
		}

		if strings.HasPrefix(trimmed, "# ") && !strings.HasPrefix(trimmed, "# Business Model Canvas") {
			if current != nil && current.Title != "" {
				canvases = append(canvases, *current)
			}
			break
		}

		if strings.HasPrefix(trimmed, "## ") {
			if current != nil && current.Title != "" {
				canvases = append(canvases, *current)
			}
			current = &domain.BusinessModelCanvas{
				ID:                    generateID(),
				Title:                 strings.TrimPrefix(trimmed, "## "),
				Date:                  time.Now().Format("2006-01-02"),
				KeyPartners:           []string{},
				KeyActivities:         []string{},
				KeyResources:          []string{},
				ValueProposition:      []string{},
				CustomerRelationships: []string{},
				Channels:              []string{},
				CustomerSegments:      []string{},
				CostStructure:         []string{},
				RevenueStreams:        []string{},
			}
			currentSubsection = ""
			continue
		}

		if current == nil {
			continue
		}

		if strings.HasPrefix(trimmed, "Date:") {
			current.Date = strings.TrimSpace(strings.TrimPrefix(trimmed, "Date:"))
			continue
		}

		if strings.HasPrefix(trimmed, "<!-- id:") {
			if match := p.idPattern.FindStringSubmatch(trimmed); len(match) > 1 {
				current.ID = match[1]
			}
			continue
		}

		// Check for section headers
		for header, key := range p.sectionMap {
			if trimmed == header {
				currentSubsection = key
				break
			}
		}

		if strings.HasPrefix(trimmed, "- ") && currentSubsection != "" {
			item := strings.TrimPrefix(trimmed, "- ")
			if item != "" {
				p.addToSection(current, currentSubsection, item)
			}
		}
	}

	if current != nil && current.Title != "" {
		canvases = append(canvases, *current)
	}

	return canvases
}

func (p *BusinessModelParser) addToSection(canvas *domain.BusinessModelCanvas, section, item string) {
	switch section {
	case "keyPartners":
		canvas.KeyPartners = append(canvas.KeyPartners, item)
	case "keyActivities":
		canvas.KeyActivities = append(canvas.KeyActivities, item)
	case "keyResources":
		canvas.KeyResources = append(canvas.KeyResources, item)
	case "valueProposition":
		canvas.ValueProposition = append(canvas.ValueProposition, item)
	case "customerRelationships":
		canvas.CustomerRelationships = append(canvas.CustomerRelationships, item)
	case "channels":
		canvas.Channels = append(canvas.Channels, item)
	case "customerSegments":
		canvas.CustomerSegments = append(canvas.CustomerSegments, item)
	case "costStructure":
		canvas.CostStructure = append(canvas.CostStructure, item)
	case "revenueStreams":
		canvas.RevenueStreams = append(canvas.RevenueStreams, item)
	}
}

// Serialize converts Business Model Canvases to markdown
func (p *BusinessModelParser) Serialize(canvases []domain.BusinessModelCanvas) string {
	var sb strings.Builder

	sb.WriteString("<!-- Business Model Canvas -->\n")
	sb.WriteString("# Business Model Canvas\n\n")

	sections := []struct {
		header string
		getter func(*domain.BusinessModelCanvas) []string
	}{
		{"Key Partners", func(c *domain.BusinessModelCanvas) []string { return c.KeyPartners }},
		{"Key Activities", func(c *domain.BusinessModelCanvas) []string { return c.KeyActivities }},
		{"Key Resources", func(c *domain.BusinessModelCanvas) []string { return c.KeyResources }},
		{"Value Proposition", func(c *domain.BusinessModelCanvas) []string { return c.ValueProposition }},
		{"Customer Relationships", func(c *domain.BusinessModelCanvas) []string { return c.CustomerRelationships }},
		{"Channels", func(c *domain.BusinessModelCanvas) []string { return c.Channels }},
		{"Customer Segments", func(c *domain.BusinessModelCanvas) []string { return c.CustomerSegments }},
		{"Cost Structure", func(c *domain.BusinessModelCanvas) []string { return c.CostStructure }},
		{"Revenue Streams", func(c *domain.BusinessModelCanvas) []string { return c.RevenueStreams }},
	}

	for _, canvas := range canvases {
		sb.WriteString(fmt.Sprintf("## %s\n", canvas.Title))
		sb.WriteString(fmt.Sprintf("<!-- id: %s -->\n", canvas.ID))
		sb.WriteString(fmt.Sprintf("Date: %s\n\n", canvas.Date))

		for _, sec := range sections {
			sb.WriteString(fmt.Sprintf("### %s\n", sec.header))
			for _, item := range sec.getter(&canvas) {
				sb.WriteString(fmt.Sprintf("- %s\n", item))
			}
			sb.WriteString("\n")
		}
	}

	return sb.String()
}

// FindByID finds a Business Model Canvas by ID
func (p *BusinessModelParser) FindByID(canvases []domain.BusinessModelCanvas, id string) *domain.BusinessModelCanvas {
	for i := range canvases {
		if canvases[i].ID == id {
			return &canvases[i]
		}
	}
	return nil
}

// GenerateID creates a unique Business Model Canvas ID
func (p *BusinessModelParser) GenerateID() string {
	return generateID()
}

File: src/lib/parser/directory/business-model.ts
/**
 * Directory-based parser for Business Model Canvas.
 * Each business model canvas is stored as a separate markdown file.
 */
import { buildFileContent, DirectoryParser, parseFrontmatter } from "./base.ts";
import type { BusinessModelCanvas } from "../../types.ts";

interface BusinessModelFrontmatter {
  id: string;
  date: string;
}

type BusinessModelSection = keyof Omit<
  BusinessModelCanvas,
  "id" | "title" | "date"
>;

const SECTION_HEADERS: Record<BusinessModelSection, string[]> = {
  keyPartners: ["key partner", "partners"],
  keyActivities: ["key activit", "activities"],
  keyResources: ["key resource", "resources"],
  valueProposition: ["value proposition"],
  customerRelationships: ["customer relationship"],
  channels: ["channel"],
  customerSegments: ["customer segment"],
  costStructure: ["cost structure", "cost"],
  revenueStreams: ["revenue stream", "revenue"],
};

export class BusinessModelDirectoryParser
  extends DirectoryParser<BusinessModelCanvas> {
  constructor(projectDir: string) {
    super({ projectDir, sectionName: "businessmodel" });
  }

  protected parseFile(
    content: string,
    _filePath: string,
  ): BusinessModelCanvas | null {
    const { frontmatter, content: body } = parseFrontmatter<
      BusinessModelFrontmatter
    >(content);

    if (!frontmatter.id) {
      return null;
    }

    const lines = body.split("\n");
    const result: BusinessModelCanvas = {
      id: frontmatter.id,
      title: "Untitled Business Model",
      date: frontmatter.date || new Date().toISOString().split("T")[0],
      keyPartners: [],
      keyActivities: [],
      keyResources: [],
      valueProposition: [],
      customerRelationships: [],
      channels: [],
      customerSegments: [],
      costStructure: [],
      revenueStreams: [],
    };

    let currentSection: BusinessModelSection | null = null;

    for (const line of lines) {
      if (line.startsWith("# ")) {
        result.title = line.slice(2).trim();
        continue;
      }

      if (line.startsWith("## ")) {
        const headerText = line.slice(3).trim().toLowerCase();
        currentSection = null;

        for (const [section, keywords] of Object.entries(SECTION_HEADERS)) {
          if (keywords.some((kw) => headerText.includes(kw))) {
            currentSection = section as BusinessModelSection;
            break;
          }
        }
        continue;
      }

      const listMatch = line.match(/^[-*]\s+(.+)$/);
      if (listMatch && currentSection) {
        result[currentSection].push(listMatch[1].trim());
      }
    }

    return result;
  }

  protected serializeItem(canvas: BusinessModelCanvas): string {
    const frontmatter: BusinessModelFrontmatter = {
      id: canvas.id,
      date: canvas.date,
    };

    const sections: string[] = [`# ${canvas.title}`];

    const addSection = (header: string, items: string[]) => {
      sections.push("");
      sections.push(`## ${header}`);
      sections.push("");
      for (const item of items) {
        sections.push(`- ${item}`);
      }
    };

    addSection("Key Partners", canvas.keyPartners);
    addSection("Key Activities", canvas.keyActivities);
    addSection("Key Resources", canvas.keyResources);
    addSection("Value Proposition", canvas.valueProposition);
    addSection("Customer Relationships", canvas.customerRelationships);
    addSection("Channels", canvas.channels);
    addSection("Customer Segments", canvas.customerSegments);
    addSection("Cost Structure", canvas.costStructure);
    addSection("Revenue Streams", canvas.revenueStreams);

    return buildFileContent(frontmatter, sections.join("\n"));
  }

  async add(
    canvas: Omit<BusinessModelCanvas, "id">,
  ): Promise<BusinessModelCanvas> {
    const newCanvas: BusinessModelCanvas = {
      ...canvas,
      id: this.generateId("bmc"),
    };
    await this.write(newCanvas);
    return newCanvas;
  }

  async update(
    id: string,
    updates: Partial<BusinessModelCanvas>,
  ): Promise<BusinessModelCanvas | null> {
    const existing = await this.read(id);
    if (!existing) return null;

    const updated: BusinessModelCanvas = {
      ...existing,
      ...updates,
      id: existing.id,
    };
    await this.write(updated);
    return updated;
  }

  async addItem(
    id: string,
    section: BusinessModelSection,
    item: string,
  ): Promise<BusinessModelCanvas | null> {
    const canvas = await this.read(id);
    if (!canvas) return null;

    canvas[section].push(item);
    await this.write(canvas);
    return canvas;
  }

  async removeItem(
    id: string,
    section: BusinessModelSection,
    itemIndex: number,
  ): Promise<BusinessModelCanvas | null> {
    const canvas = await this.read(id);
    if (!canvas) return null;

    if (itemIndex >= 0 && itemIndex < canvas[section].length) {
      canvas[section].splice(itemIndex, 1);
      await this.write(canvas);
    }
    return canvas;
  }
}
